flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/FunctionCatalog.java

```diff
@@ -136,18 +185,19 @@ public void registerScalarFunction(String name, ScalarFunction function) {
 
 			foundDefinition = BuiltInFunctionDefinitions.getDefinitions()
 				.stream()
-				.filter(f -> normalizeName(name).equals(normalizeName(f.getName())))
+				.filter(f -> functionName.equals(normalizeName(f.getName())))
 				.findFirst()
 				.map(Function.identity());
 		}
 
 		return foundDefinition.map(definition -> new FunctionLookup.Result(
-			ObjectIdentifier.of(defaultCatalogName, defaultDatabaseName, name),
+			ObjectIdentifier.of(catalogManager.getCurrentCatalog(), catalogManager.getCurrentDatabase(), name),
```

Comment:
I'm a little confused/concerned about the identifier created here. From about code it seems that the function can be resolved either in the current catalog (in catalog manager), in "userFunctions", which might conceptually belong to default catalog/database (or built-in catalog), or in BuiltInFunctionDefinitions which doesn't belong to any catalog. As a followup, we might want to unify userFunctions with built-in catalog and clarify the resolution process in multiple catalogs.

Reply:
yes, that's a very good point. I will revert the path of in-memory and built-in functions to its original behavior and follow up to see how to deal with path of built-in functions.

Reply:
Unifying built-in functions and catalog functions is on our roadmap for 1.10. FunctionCatalog will not be necessary then.